File: 2/client/main.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
	"time"
)

type Config struct {
	ServerAddress string `json:"server_address"`
	ServerPort    int    `json:"server_port"`
	SendInterval  int    `json:"send_interval"`
	Name          string `json:"name"`
	Group         string `json:"group"`
	LogFile       string `json:"log_file"`
}

type LogEntry struct {
	Timestamp     time.Time `json:"timestamp"`
	EventType     string    `json:"event_type"`
	Details       string    `json:"details"`
	ServerAddress string    `json:"server_address"`
}

func main() {
	config, err := loadConfig("./2/client/config.json")
	if err != nil {
		log.Fatalf("Ошибка при чтении конфигурации: %v", err)
	}

	logFile, err := os.OpenFile(config.LogFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatalf("Ошибка при открытии лог-файла: %v", err)
	}
	defer logFile.Close()

	logger := log.New(logFile, "CLIENT: ", log.Ldate|log.Ltime|log.Lshortfile)

	serverAddress := fmt.Sprintf("%s:%d", config.ServerAddress, config.ServerPort)
	conn, err := net.Dial("tcp", serverAddress)
	if err != nil {
		logger.Fatalf("Ошибка при подключении к серверу %s: %v", serverAddress, err)
	}
	defer conn.Close()

	logEvent(logger, "connect", "Успешное подключение", config.ServerAddress)

	ticker := time.NewTicker(time.Duration(config.SendInterval) * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		message := fmt.Sprintf("%s %s", config.Name, config.Group)

		_, err := fmt.Fprintf(conn, message+"\n")
		if err != nil {
			logEvent(logger, "error", fmt.Sprintf("Ошибка при отправке сообщения: %v", err), config.ServerAddress)
			continue
		}
		logEvent(logger, "send", fmt.Sprintf("Отправлено сообщение: %s", message), config.ServerAddress)

		conn.SetReadDeadline(time.Now().Add(10 * time.Second))
		response, err := bufio.NewReader(conn).ReadString('\n')
		if err != nil {
			logEvent(logger, "error", fmt.Sprintf("Ошибка при получении ответа: %v", err), config.ServerAddress)
			continue
		}
		logEvent(logger, "receive", fmt.Sprintf("Получено сообщение: %s", strings.TrimSpace(response)), config.ServerAddress)
	}
}

func loadConfig(filename string) (Config, error) {
	var config Config
	file, err := os.Open(filename)
	if err != nil {
		return config, err
	}
	defer file.Close()

	decoder := json.NewDecoder(file)
	err = decoder.Decode(&config)
	if err != nil {
		return config, err
	}

	return config, nil
}

func logEvent(logger *log.Logger, eventType string, details string, serverAddress string) {
	entry := LogEntry{
		Timestamp:     time.Now(),
		EventType:     eventType,
		Details:       details,
		ServerAddress: serverAddress,
	}

	jsonEntry, err := json.Marshal(entry)
	if err != nil {
		logger.Printf("Ошибка при сериализации в JSON: %v", err)
		logger.Printf("%s: %s - %s - %s", entry.Timestamp.Format(time.RFC3339), eventType, details, serverAddress)
		return
	}

	logger.Printf("%s\n", string(jsonEntry))
}

File: 2/python/client/main.py
import json
import socket
import time
import logging

def load_config(filename):
    with open(filename, "r", encoding="utf-8") as f:
        return json.load(f)

def log_event(logger, event_type, details, server_address):
    entry = {
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S", time.localtime()),
        "event_type": event_type,
        "details": details,
        "server_address": server_address
    }
    try:
        json_entry = json.dumps(entry, ensure_ascii=False)
        logger.info(json_entry)
    except Exception as e:
        logger.error("Ошибка при сериализации в JSON: %s", e)
        logger.info(f"{entry['timestamp']}: {event_type} - {details} - {server_address}")

def main():
    config_path = "config.json"
    try:
        config = load_config(config_path)
    except Exception as e:
        print(f"Ошибка при чтении конфигурации: {e}")
        return

    log_file = config.get("log_file", "client.log")
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s %(levelname)s: %(message)s",
        handlers=[logging.FileHandler(log_file, encoding="utf-8", mode="a")]
    )
    logger = logging.getLogger("CLIENT")

    server_address = config["server_address"]
    server_port = config["server_port"]

    try:
        sock = socket.create_connection((server_address, server_port))
    except Exception as e:
        logger.error("Ошибка при подключении к серверу %s:%s: %s", server_address, server_port, e)
        return

    log_event(logger, "connect", "Успешное подключение", server_address)
    send_interval = config["send_interval"]

    while True:
        message = f"{config['name']} {config['group']}\n"
        logger.info("Отправляем: " + message.strip())
        try:
            sock.sendall(message.encode('utf-8'))
            log_event(logger, "send", f"Отправлено сообщение: {message.strip()}", server_address)
        except Exception as e:
            log_event(logger, "error", f"Ошибка при отправке сообщения: {e}", server_address)
            time.sleep(send_interval)
            continue

        try:
            response_bytes = b""
            while not response_bytes.endswith(b"\n"):
                chunk = sock.recv(1024)
                if not chunk:
                    raise Exception("Нет ответа от сервера")
                response_bytes += chunk
            response = response_bytes.decode('utf-8').strip()
            log_event(logger, "receive", f"Получено сообщение: {response}", server_address)
        except Exception as e:
            log_event(logger, "error", f"Ошибка при получении ответа: {e}", server_address)
        time.sleep(send_interval)

if __name__ == "__main__":
    main()
